### DISCUSSION TITLE
Stash a cc65 program and reload it

### DISCUSSION THREAD

**Discussion by cyberfoxIT** [cc65/cc65#1761] [Q&A]
Hello,

I'm trying to save a program loaded from disk on my REU and reload it in a second time.
So I've made a loader that read a prg file and save it on the reu, once loaded it launches a small ml program that fetch the program from the reu and execute it.
The program itself start, but when it try to exec another pgm from it (with exec function) the disk start spin, but nothing more.
What should I care of when I stash and fetch a program to and from the REU ?
The same problem if I stash and fetch from basic.

Thank you in advance for your help!

**Comment by polluks** [comment]
But exec works without REU at all?

**Reply by cyberfoxIT** [reply]
Yes, if I load and run the program directly, yes it work as it should with or without a REU.

The problem is that when loaded from disk there's no problem, when fetched from the REU something goes wrong.
When I load on the REU I skip the first 14 bytes that should contain the basic part so I copy from 7181 (1C0D) for the length of the pgm.

**Comment by cyberfoxIT** [comment]
Yes, if I load and run the program directly, yes it work as it should.

The problem is that when loaded from disk there's no problem, when fetched from the REU something goes wrong.
When I load on the REU I skip the first 14 bytes that should contain the basic part so I copy from 7181 (1C0D) for the length of the pgm.

**Reply by spiro-trikaliotis** [reply]
What happens if you do not skip the 14 byte? Does it work then?
This would give a valuable hint on what is going on.

**Reply by cyberfoxIT** [reply]
Just the same...
The problem seem to occurs on disk access, so I thought was maybe the bank that as been changed somewhere because I've read that CC65 pgm's use a non standard configuration...
So I've set FF00 to 0x0e before jsr to the pgm, but not luck...

**Reply by spiro-trikaliotis** [reply]
So you are on a C128?

What comes to mind (note: I am using the C64 address; look up their equivalents on a C128!):

- $BA, address of the last used drive (from the last load command, for example); you might use that?
- $2D/$2E, the end of the BASIC program (and the other vectors afterwards, which can be fixed by a CLR BASIC command)

Another test to find the cause might be that you do the following:
1. Store a program (for example, only the BASIC stub) on disk
2. LOAD the correct program, stash it into the REU, and start it
3. Now, when you want to re-start it LOAD the stub program from 1.
4. Now, fetch the program from the REU
5. Fix the end of BASIC pointer (equivalent of $2D/E on the C128)
6. Re-start the program

Does that work?

It might be also some specific thing of your program, or your REU operation is done wrong, or ...

It might help if you could give a fully working example (or, more specifically, an example that does NOT work).

**Comment by greg-king5** [comment]
On CBM computers, `exec()` uses BASIC to load and run a program.  It requires that its program was started by RUNning the BASIC header.  You must find out what ROM subroutines can be used to do what the RUN statement does.

**Comment by cyberfoxIT** [comment]
I've made this simple test.
I created a pgm that exec what you write in a scanf.
In the emulator If I load the pgm and I run it, no problems at all.
If I load it and stash it with "stash 42000,dec("1c00"),0,0".
I reset the emulator, I fetch the program from the reu with "fetch 42000,dec("1c00"),0,0".
At this point I can see the list of the basic pgm, but if I run it with "run" it goes well, if I run it doing "sys 7181", that is what the basic program does, it doesn't work.
Where's the difference between running it with "run" instead of "sys7181" ? This is what's missing in my opinion.

**Comment by cyberfoxIT** [comment]
Solved!
Instead of sys it I invoke the run statement from my lm, and like this it's working fine...
